futurex_openedx_extensions/dashboard/serializers.py

```diff
@@ -25,60 +33,155 @@ class LearnerDetailsSerializer(serializers.ModelSerializer):
     class Meta:
         model = get_user_model()
         fields = [
-            'user_id',
-            'full_name',
-            'username',
-            'email',
-            'mobile_no',
-            'year_of_birth',
-            'gender',
-            'date_joined',
-            'last_login',
-            'enrolled_courses_count',
-            'certificates_count',
+            "user_id",
+            "full_name",
+            "alternative_full_name",
+            "username",
+            "email",
+            "mobile_no",
+            "year_of_birth",
+            "gender",
+            "gender_display",
+            "date_joined",
+            "last_login",
+            "enrolled_courses_count",
+            "certificates_count",
         ]
 
+    def _get_names(self, obj, alternative=False):
+        """
+        Calculate the full name and alternative full name. We have two issues in the data:
+        1. The first and last names in auth.user contain many records with identical values (redundant data).
+        2. The name field in the profile sometimes contains data while the first and last names are empty.
+        """
+        first_name = obj.first_name.strip()
+        last_name = obj.last_name.strip()
+        alt_name = (self._get_profile_field(obj, "name") or "").strip()
+
+        if not last_name:
+            full_name = first_name
+        elif not first_name:
+            full_name = last_name
+        elif first_name == last_name and " " in first_name:
+            full_name = first_name
+        else:
+            full_name = f"{first_name} {last_name}"
+
+        if alt_name == full_name:
+            alt_name = ""
+
+        if not full_name and alt_name:
+            full_name = alt_name
+            alt_name = ""
+
+        if alt_name and ord(alt_name[0]) > 127 >= ord(full_name[0]):
+            names = alt_name, full_name
+        else:
+            names = full_name, alt_name
+
+        return names[0] if not alternative else names[1]
+
```

Comment:
This code could benefit from refactoring to simplify its complexity. I also prepared a secret [gist](https://gist.github.com/iamjazzar/fd62c265a63af1f3293fea0a0732c4c2) containing the tests I've developed to ensure the end results align as expected. I'd also rename the function since it returns one name.

```python
    def _get_name(self, obj, alternative=False):
        """
        Calculate the full name and alternative full name. We have two issues in the data:
        1. The first and last names in auth.user contain many records with identical values (redundant data).
        2. The name field in the profile sometimes contains data while the first and last names are empty.
        """
        first_name = obj.first_name.strip()
        last_name = obj.last_name.strip()
        full_name = " ".join(filter(None, (first_name, last_name)))

        profile_name = self._get_profile_field(obj, "name")
        alt_name = profile_name.strip() if profile_name else ""

        # If full name is empty then use Alt name
        if not full_name:
            full_name, alt_name = alt_name, ""

        # Swap names if alt_name is not in English
        if alt_name and not is_english(alt_name):
            full_name, alt_name = alt_name, full_name

        # Return full_name or alt_name based on the 'alternative' flag
        return alt_name if alternative else full_name
```

And the English helper function can look like this:
```python
def is_english(text):
    """
    Checks if a string consists only of characters in the ASCII range (a-z, A-Z, 0-9, and common symbols).

    This method is very basic and may miss valid English text with non-ASCII characters.
    """
    return all(ord(char) < 128 for char in text)
```

Reply:
Thank you for the effort enhancing that! I've adopted the refactor with a small change. 

```python
- full_name = " ".join(filter(None, (first_name, last_name)))
+ full_name = first_name or last_name
+   if first_name and last_name and not (first_name == last_name and " " in first_name):
+     full_name = " ".join(filter(None, (first_name, last_name)))
```

it's because I want to eliminate the view of some bad data. We have too many records having `first_name` and `last_name` containing a Full-Name !

Reply:
Looks good.